### ISSUE TITLE
Should a casher timeout failure clear cached directories?

### ISSUE THREAD

**Issue by tkelman** [travis-ci/travis-ci#5329]
Earlier today there were a few builds on JuliaLang/julia that hit 3-minute timeouts when trying to load their cache: https://travis-ci.org/JuliaLang/julia/jobs/97551862

relevant portion of log:
```
Setting up build cache
$ export CASHER_DIR=$HOME/.casher
$ Installing caching utilities
attempting to download cache archive
fetching master/cache--compiler-gpp-m32.tgz
running `casher fetch` took longer than 180 seconds and has been aborted
adding /home/travis/build/JuliaLang/julia/deps/srccache to cache
adding /home/travis/build/JuliaLang/julia/deps/build to cache
```

Later on, the build failed with a `Makefile:44: Makefile.config: No such file or directory` error that I've never seen before this, and is not at all related to the commit that was being built there (it didn't touch makefiles, only 2 `.jl` files that do not get built until much later in a successful job). It seems as though the cache may have been partially extracted and hence in a corrupt state, instead of trying to rebuild the contents of the cache from scratch.

Several more builds after this failed as well with the same error until we noticed the issue, manually cleared the cache and restarted a build on master to re-populate the cache from scratch. Would be good to put some mitigation in place so this doesn't happen again next time, either on Travis' side or manually in our yml. Maybe we could post a checksum of the cache contents from every successful build to somewhere that we would check against ahead of time? Seems like others might hit this though so may be good to put this kind of consistency check or fallback somewhere it would get used more broadly. Is deleting the cache in case of casher timeout an appropriate general-purpose fallback, no cache (in case of network issues, hopefully rare) better than possibly corrupt partial cache?

**Comment by tkelman** [comment]
Or maybe easier - a failed build shouldn't be allowed to update the cache?

**Comment by BanzaiMan** [comment]
@tkelman Failed jobs uploading the cache seems like another issue that should be dealt with separately.

**Comment by tkelman** [comment]
Indeed, it already is open as https://github.com/travis-ci/travis-ci/issues/4472 - sounds like it should be configurable though since some people want failed builds to still be able to do the initial populating of the cache. In the case of a timeout that seems like it could be worked around, e.g. don't run tests if cache content does not exist yet.

**Comment by BanzaiMan** [comment]
I spotted a bug in the caching utility. I plan to work on that in the new year. Thanks for your patience.
